### ISSUE TITLE
Gemsym doesn't update its counter until control reaches the top level.

### ISSUE THREAD

**Issue by sdilts** [carp-lang/Carp#696]
The code 
``` clojure
(let [a (gensym)]
  (let [b (gensym)]
    (do
      (macro-log a)
      (macro-log b))))
```
prints out 
``` clojure
gensym1000
gensym1000
```
instead of the expected
``` clojure
gensym1000
gensym1001
```
When the top level is reached, you can see that `*gensym-counter*` has the correct value of `1002`, so the counter is being incremented, but the value isn't being updated at the correct time.

 Here is the code responsible:
https://github.com/carp-lang/Carp/blob/85eacb0ebf68b99f9ae72539e7cd2a0c9e31f02f/core/Macros.carp#L357-L368

**Comment by hellerve** [comment]
I’ve seen this happen as well, and Im pretty sure this will eventually be fixed in #695. I‘m pretty sure the semantics of `do` are at fault here.

**Comment by hellerve** [comment]
Eyeballing [`do`](https://github.com/carp-lang/Carp/blob/master/src/Eval.hs#L98), it definitely looks like the same problem that we had to fix with `let` in #678 to make bindings depend on each other.

This doesn’t explain why the counter doesn’t update between the two calls to `let`, though, and in fact combining the two `let`s into one doesn’t help either. This tells me that there is something else at play that I‘m just not seeing yet.

**Comment by hellerve** [comment]
This can be closed now!
